browser/fxr/content/permissions.js

```diff
@@ -42,65 +50,94 @@ class FxrPermissionPromptPrototype {
       this.deny();
     }
 
-    this.responseCallback();
+    if (this.ui) {
+      this.ui.close();
+      this.ui = null;
+    }
+
+    let status = {};
+    status[allowed ? "allowed" : "denied"] = this.promptTypes;
+    this.responseCallback(status);
   }
 }
 
 // WebRTC-specific class implementation
+//
+// For now-- only supports microphone permissions with the following behavior:
+// - per domain/principal, allowed is required for every request
+// - per domain/principal, denied is persisted for the session (i.e., must
+//   restart the browser )
+// The UI behavior to support this also includes code in WebRTCParent.jsm and
+// fxrui.js.
 class FxrWebRTCPrompt extends FxrPermissionPromptPrototype {
-  showPrompt() {
-    for (let typeName of this.request.requestTypes) {
-      if (typeName !== "Microphone" && typeName !== "Camera") {
-        // Only Microphone and Camera requests are allowed. Automatically deny
-        // any other request.
-        this.defaultDeny();
-        return;
-      }
-    }
-
-    super.showPrompt();
-  }
+  constructor(aRequest, aBrowser, aCallback) {
+    super(aRequest, aBrowser, aCallback);
 
-  allow() {
     let { audioDevices, videoDevices } = this.request;
 
-    let principal = Services.scriptSecurityManager.createContentPrincipalFromOrigin(
+    this.principal = Services.scriptSecurityManager.createContentPrincipalFromOrigin(
       this.request.origin
     );
 
     // For now, collect the first audio and video device by default. User
     // selection will be enabled later:
     // Bug 1594841 - Add UI to select device for WebRTC in FxR for Desktop
-    let allowedDevices = [];
+    this.promptedDevices = [];
 
     if (audioDevices.length) {
-      allowedDevices.push(audioDevices[0].deviceIndex);
+      this.promptedDevices.push(audioDevices[0]);
     }
+  }
 
-    if (videoDevices.length) {
-      Services.perms.addFromPrincipal(
-        principal,
-        "MediaManagerVideo",
-        Services.perms.ALLOW_ACTION,
-        Services.perms.EXPIRE_SESSION
-      );
-      allowedDevices.push(videoDevices[0].deviceIndex);
+  showPrompt() {
+    if (this.request.requestTypes.length === 1
+      && this.request.requestTypes[0] === "Microphone") {
+      // Only Microphone requests are allowed. Automatically deny any other
+      // request.
+      this.promptTypes = this.request.requestTypes;
+      this.ui = new FxrPermissionUI(
+        this,
+        this.promptTypes,
+        this.promptedDevices,
+        this.request.origin
+        );
+
+      return;
+    } else {
+      // Otherwise, proceed to deny by default
+      super.showPrompt();
     }
+  }
 
+  allow() {
     // WebRTCChild doesn't currently care which actor
     // this is sent to and just uses the windowID.
+    let deviceIndices = [];
+    for (let i = 0; i < this.promptedDevices.length; i++) {
```

Comment:
nit: Could use map here
`let deviceIndices = this.promptedDevices.map((device) => device.deviceIndex))`